### PR TITLE
Resolve vpc value prior to checking it's value

### LINKED ISSUE #6327: Unable to deploy with VPC from external shared project
### [REQUIRED] Environment info

**firebase-tools:** 12.5.2

**Platform:** macos

### [REQUIRED] Test case

Trying to use a shared vpc from a different project with `.env` and

```
FUNCTION_VPC="projects/shared/locations/us-central1/connectors/shared"
```

```js
const vpc = params.defineString('FUNCTION_VPC');
setGlobalOptions({
  vpcConnector: vpc
});
```

### [REQUIRED] Steps to reproduce

Try to deploy with these values and firebase tools mutates the `vpcConnector` string to an invalid value.

```
functions: HTTP Error: 400, Could not update Cloud Run service projects/my-project/locations/us-central1/services/function. spec.template.metadata.annotations: Expected a VPC access connector name like projects/*/locations/*/connectors/*, but obtained projects/not-shared-project/locations/us-central1/connectors/{{the full value of the FUNCTION_VPC}}
```

I assume the issue is around https://github.com/firebase/firebase-tools/blob/09c8d88ff5f1b4c4e617064c7da729b2301d905b/src/deploy/functions/runtimes/node/parseTriggers.ts#L502-L514

But my connector has a `/` in it so it shouldn't get modified. I wonder if it has to do with the `Expression<string>` type that has a `.value()` not being invoked or something until later? Or maybe the `/` is encoded when read from the environment file?

### [REQUIRED] Expected behavior

The VPC value I specify is used

### [REQUIRED] Actual behavior

The string is mutated and the deployment fails.

### PR DESCRIPTION
**Pull request by steveoh**
### Description

fixes #6327

### CODE REVIEW

**Comment by steveoh** [comment]
I'd love to got a workflow approval for this to see the tests run!

**Comment by steveoh** [comment]
@joehan 🔔

**Comment by joehan** [comment]
Hey @steveoh, thanks for the @ on this, it managed to get lost in the mix. I've approved the test run and it looks like there is a unit test case failing.

I've also assigned some folks who work on CF3 to review.

**Comment by steveoh** [comment]
Thanks @joehan. 

For time savings this is the failed test

```
1) addResourcesToBuild
       should expand vpc connector in w/ shorthand form:

      AssertionError: expected { Object (requiredAPIs, endpoints, ...) } to deeply equal { Object (requiredAPIs, endpoints, ...) }
      + expected - actual

               "project": "project"
               "region": "us-central1"
               "runtime": "nodejs16"
               "vpc": {
      -          "connector": "hello-vpc"
      +          "connector": "projects/project/locations/us-central1/connectors/hello-vpc"
               }
             }
           }
         }
```

I think this test is wrong and should be updated. The values are equivalent basically and the long form could be seen as more correct™ as it allows for more use cases? Thoughts?

File: src/deploy/functions/build.ts [hunk 1]
```diff
@@ -494,10 +494,10 @@ export function toBackend(
         nullsafeVisitor((cpu) => (cpu === "gcf_gen1" ? cpu : r.resolveInt(cpu)))
       );
       if (bdEndpoint.vpc) {
+        bkEndpoint.vpc = { connector: params.resolveString(bdEndpoint.vpc.connector, paramValues) };
```
Comment (colerogers) [inline src/deploy/functions/build.ts:497]: You want to resolve the `bdEndpoint.vpc.connector` and save that back into `bdEndpoint` so the `if` check below works. Something like this should do it:
```
if (bdEndpoint.vpc) {
        **bdEndpoint.vpc.connector = params.resolveString(bdEndpoint.vpc.connector, paramValues);**
        if (bdEndpoint.vpc.connector && !bdEndpoint.vpc.connector.includes("/")) {
          bdEndpoint.vpc.connector = `projects/${bdEndpoint.project}/locations/${region}/connectors/${bdEndpoint.vpc.connector}`;
        }
        **bkEndpoint.vpc = { connector: bdEndpoint.vpc.connector };**
        proto.copyIfPresent(bkEndpoint.vpc, bdEndpoint.vpc, "egressSettings");
      } ...
```

**Review by colerogers — COMMENTED** [review]
small bug, please fix and rerun the tests

**Comment by colerogers** [comment]
> Thanks @joehan.
> 
> For time savings this is the failed test
> 
> ```diff
> 1) addResourcesToBuild
>        should expand vpc connector in w/ shorthand form:
> 
>       AssertionError: expected { Object (requiredAPIs, endpoints, ...) } to deeply equal { Object (requiredAPIs, endpoints, ...) }
>       + expected - actual
> 
>                "project": "project"
>                "region": "us-central1"
>                "runtime": "nodejs16"
>                "vpc": {
>       -          "connector": "hello-vpc"
>       +          "connector": "projects/project/locations/us-central1/connectors/hello-vpc"
>                }
>              }
>            }
>          }
> ```
> 
> I think this test is wrong and should be updated. The values are equivalent, as long as `hello-vpc` is in that project and region, and the long form could be seen as more correct™ as it allows for the shared vpc use case? Thoughts?

If you change your code to save to `bdEndpoint` and not `bkEndpoint`, the test should pass

**Comment by steveoh** [comment]
@colerogers does there need to be another test case written for this use case?

File: src/deploy/functions/build.ts (hunk 1 above)
Comment (steveoh) [reply src/deploy/functions/build.ts]: thanks!

**Comment by steveoh** [comment]
@joehan can we approve the test runs?

**Comment by steveoh** [comment]
ughhh, the codespace wasn't set up with prettier...

**Comment by colerogers** [comment]
@steveoh please run `npm run mocha` to test everything out locally

File: src/deploy/functions/build.ts [hunk 2]
```diff
@@ -494,10 +494,12 @@ export function toBackend(
         nullsafeVisitor((cpu) => (cpu === "gcf_gen1" ? cpu : r.resolveInt(cpu)))
       );
       if (bdEndpoint.vpc) {
+        bdEndpoint.vpc = { connector: params.resolveString(bdEndpoint.vpc.connector, paramValues) };
```
Comment (colerogers) [inline src/deploy/functions/build.ts:497]: You are writing over the entire `vpc` object which will remove the field `egressSettings` if it exists

Comment (colerogers) [reply src/deploy/functions/build.ts:497]:
```
bdEndpoint.vpc.connector = params.resolveString(bdEndpoint.vpc.connector, paramValues);
```

Comment (steveoh) [reply src/deploy/functions/build.ts:497]: ok committed that fix.

**Comment by steveoh** [comment]
> @steveoh please run `npm run mocha` to test everything out locally

ok, i never checked out the repo. I debugged in my node_modules and was hoping your ci would do the trick but it is a pain with the workflows requiring a review. I'll check it locally and test it out.

**Comment by steveoh** [comment]
ok tests passed locally @joehan

**Comment by steveoh** [comment]
ok tests passed locally. Can I get a CI review approval to cross the finish line @joehan?

dwight cracking nuckles

**Review by joehan — APPROVED** [review]

**Comment by joehan** [comment]
LGTM - will get this out in the next release later this week. Thanks for the contribution @steveoh!

**Pull request merged**